### PR TITLE
Rewriting dead branch elimination.

### PR DESCRIPTION
**Pull request by alan-baker**
Addresses #1157 

Pass now paints live blocks and fixes constant branches and switches as
it goes. No longer requires structured control flow. It also removes
unreachable blocks as a side effect. It fixes the IR (phis) before doing
any code removal (other than terminator changes).

Added several unit tests for updated/new functionality.

### CODE REVIEW

**Comment by dneto0** [comment]
Clang errors out due to unused captures.

File: test/opt/dead_branch_elim_test.cpp [hunk 1]
```diff
@@ -1334,7 +1334,7 @@ OpBranch %33
 OpStore %oc %19
 OpBranch %28
 %29 = OpLabel
-OpBranchConditional %false %27 %28
+OpBranch %27
```
Comment (dneto0) [inline test/opt/dead_branch_elim_test.cpp]: There is no longer a backedge for this loop.  That's invalid.

Comment (alan-baker) [reply test/opt/dead_branch_elim_test.cpp]: This branch is the backedge (%27 contains OpLoopMerge). The exit to the merge is no longer present.

Comment (dneto0) [reply test/opt/dead_branch_elim_test.cpp]: whoops.  You're right.

**Comment by greg-lunarg** [comment]
VS does not like the trailing comma after liveLabId. :(

**Comment by dneto0** [comment]
> VS does not like the trailing comma after liveLabId. :(

And it shows up as an error or warning in YouCompleteMe with my GCC.

**Comment by dneto0** [comment]
While we're here, @greg-lunarg : This rewrite handles some extra cases that @alan-baker is seeing.
  
An example is where there's a selection construct but control flow reconverges early, one block before the merge block, and that early block has a Phi.  The old algorithm doesn't know to fixup that phi because it never sees it.  The result with the old algorithm is a phi with dangling references to the dead block that has been removed.  See the EarlyReconvergence test
https://github.com/KhronosGroup/SPIRV-Tools/pull/1173/files#diff-34499625bbeb0676d52a53a9454d9e96R1461

File: source/opt/dead_branch_elim_pass.cpp [hunk 2]
```diff
@@ -73,260 +72,215 @@ bool DeadBranchElimPass::GetConstInteger(uint32_t selId, uint32_t* selVal) {
...
-  for (auto bi = structuredOrder.begin(); bi != structuredOrder.end(); ++bi) {
-    // Skip blocks that are already in the elimination set
-    if (elimBlocks.find(*bi) != elimBlocks.end()) continue;
-    // Skip blocks that don't have conditional branch preceded
-    // by OpSelectionMerge
-    ir::Instruction* br;
-    ir::Instruction* mergeInst;
-    uint32_t condId;
-    if (!GetSelectionBranch(*bi, &br, &mergeInst, &condId)) continue;
+  auto get_parent_block = [this](uint32_t id) {
```
Comment (dneto0) [inline source/opt/dead_branch_elim_pass.cpp]: Should say that we only need the instr-to-block mapping in for instructions that exist when we start processing or that we explicitly analyze later. You don't need to update it for instructions that are removed.
That explains why later creation of branch instructions don't need to invalidate or update the analysis.

**Comment by greg-lunarg** [comment]
> While we're here, @greg-lunarg : This rewrite handles some extra cases that @alan-baker is seeing.

Totally cool. However, I am seeing regressions. I am tracking them down now.

File: source/opt/dead_branch_elim_pass.cpp [hunk 3]
```diff
@@ -73,260 +72,215 @@ bool DeadBranchElimPass::GetConstInteger(uint32_t selId, uint32_t* selVal) {
...
-      dLabId = (*dbi)->id();
+  // Check for unreachable merge and continue blocks with live headers, those
+  // blocks must remain. Continues are tracked separately so that when updating
+  // live phi nodes with an edge from a continue I can replace it with an
+  // undef (because I still clobber the continue block).
+  std::unordered_set<ir::BasicBlock*> unreachableMerges;
+  std::unordered_map<ir::BasicBlock*, ir::BasicBlock*> unreachableContinues;
+  for (auto block : liveBlocks) {
+    uint32_t mergeId = block->MergeBlockIdIfAny();
+    if (mergeId != 0) {
```
Comment (dneto0) [inline source/opt/dead_branch_elim_pass.cpp]: Can shorten this by one line by putting the declaration inside the "if".
```
if (auto merge_id = block->MergeBlockIdIfAny()) {
...
```

File: source/opt/dead_branch_elim_pass.cpp [hunk 4]
```diff
@@ -73,260 +72,215 @@ bool DeadBranchElimPass::GetConstInteger(uint32_t selId, uint32_t* selVal) {
...
-      elimBlocks.insert(*dbi);
-      ++dbi;
-      if (dbi == structuredOrder.end()) break;
-      dLabId = (*dbi)->id();
+  // Check for unreachable merge and continue blocks with live headers, those
+  // blocks must remain. Continues are tracked separately so that when updating
+  // live phi nodes with an edge from a continue I can replace it with an
+  // undef (because I still clobber the continue block).
+  std::unordered_set<ir::BasicBlock*> unreachableMerges;
+  std::unordered_map<ir::BasicBlock*, ir::BasicBlock*> unreachableContinues;
```
Comment (dneto0) [inline source/opt/dead_branch_elim_pass.cpp]: Describe what is mapped to what.

File: source/opt/dead_branch_elim_pass.cpp [hunk 5]
```diff
@@ -73,260 +72,215 @@ bool DeadBranchElimPass::GetConstInteger(uint32_t selId, uint32_t* selVal) {
...
-      backedges_.erase((*dbi)->terminator());
-      KillAllInsts(*dbi);
-      elimBlocks.insert(*dbi);
-      ++dbi;
-      if (dbi == structuredOrder.end()) break;
-      dLabId = (*dbi)->id();
+  // Check for unreachable merge and continue blocks with live headers, those
+  // blocks must remain. Continues are tracked separately so that when updating
+  // live phi nodes with an edge from a continue I can replace it with an
+  // undef (because I still clobber the continue block).
```
Comment (dneto0) [inline source/opt/dead_branch_elim_pass.cpp]: "(because we clobber the instructions inside of the continue block)."  

Most important is the clarification that the definition instructions inside the continue block are removed, but the continue block itself survives.

The "still" wasn't doing anything for you here.  And I use the royal "we" because the reader should feel they own the code too.

File: source/opt/dead_branch_elim_pass.cpp [hunk 6]
```diff
@@ -73,260 +72,215 @@ bool DeadBranchElimPass::GetConstInteger(uint32_t selId, uint32_t* selVal) {
...
+        if (!liveBlocks.count(contBlock)) {
+          unreachableContinues[contBlock] = block;
+        }
+      }
     }
+  }
 
-    // If last block reached, look for next dead branch
-    if (dbi == structuredOrder.end()) continue;
+  for (auto& block : *func) {
```
Comment (dneto0) [inline source/opt/dead_branch_elim_pass.cpp]: Given the length of this function, you need a comment saying what this phase does.    I guess I suggest moving the comment starting with "Fix phis" out to the above the iteration over the blocks.

Then rehprase it as   "Fix phis in reachable blocks so that ...."

File: source/opt/dead_branch_elim_pass.cpp [hunk 7]
```diff
@@ -73,260 +72,215 @@ bool DeadBranchElimPass::GetConstInteger(uint32_t selId, uint32_t* selVal) {
...
-          if (deadPreds.find(cfg()->block(*idp)) == deadPreds.end()) {
-            ++lcnt;
-            lidx = icnt - 1;
+        if (changed) {
+          // Replace the phi with either a single value or a rebuilt phi.
+          uint32_t replId;
+          if (operands.size() == 2) {
+            replId = operands[0].words[0];
+          } else {
+            replId = TakeNextId();
```
Comment (dneto0) [inline source/opt/dead_branch_elim_pass.cpp]: If we're going to delete the old phi we should reuse its result id, at least in principle.
  
That makes it tricky to keep inst-def-use up to date but it should be doable.  You'd have to kill the old instruction then add the new one.

File: source/opt/dead_branch_elim_pass.cpp [hunk 8]
```diff
@@ -73,260 +72,215 @@ bool DeadBranchElimPass::GetConstInteger(uint32_t selId, uint32_t* selVal) {
...
     }
   }
 
-  // Erase dead blocks
-  for (auto ebi = func->begin(); ebi != func->end();)
-    if (elimBlocks.find(&*ebi) != elimBlocks.end())
+  // Erase dead blocks.
+  for (auto ebi = func->begin(); ebi != func->end();) {
+    if (unreachableMerges.count(&*ebi)) {
+      // Make unreachable, but leave the label.
```
Comment (dneto0) [inline source/opt/dead_branch_elim_pass.cpp]: Might be worth saying the unreachable merge and continue blocks need to remain they can be targeted by a merge instruction.
Any other use of the label is, perforce, a branch from yet another unreachable block.  And that other unreachable block will itself either (a) disappear or (b) be emptied out and its terminator replaced with an OpUnreachable or a trivial backedge.

**Comment by greg-lunarg** [comment]
Here is a pre-packageed test with a fairly simple piece of code for which this PR currently generates incorrect code. Essentially, it is an if-then with a phi in the merge block. This pattern is quite commonly generated by glslang for short-circuit boolean operations, although I believe the code in this test reasonably exemplifies the basic problem.

There might be other regressions as well. I will start with this and will re-run my regression testing once this problem is addressed.

TEST_F(DeadBranchElimTest, IfThenPhiTrue) {
  // Test handling of phi in merge block after dead branch elimination.
  // Note: The SPIR-V has had store/load elimination and phi insertion
  //
  // #version 140
  //
  // void main()
  // {
  //     vec4 v;
  //     v = vec4(0.0,0.0,0.0,0.0);
  //     if (true)
  //       v = vec4(1.0,1.0,1.0,1.0);
  //     gl_FragColor = v;
  // }

  const std::string predefs =
      R"(OpCapability Shader
%1 = OpExtInstImport "GLSL.std.450"
OpMemoryModel Logical GLSL450
OpEntryPoint Fragment %main "main" %gl_FragColor
OpExecutionMode %main OriginUpperLeft
OpSource GLSL 140
OpName %main "main"
OpName %gl_FragColor "gl_FragColor"
%void = OpTypeVoid
%5 = OpTypeFunction %void
%bool = OpTypeBool
%true = OpConstantTrue %bool
%float = OpTypeFloat 32
%v4float = OpTypeVector %float 4
%_ptr_Function_v4float = OpTypePointer Function %v4float
%float_0 = OpConstant %float 0
%12 = OpConstantComposite %v4float %float_0 %float_0 %float_0 %float_0
%float_1 = OpConstant %float 1
%14 = OpConstantComposite %v4float %float_1 %float_1 %float_1 %float_1
%_ptr_Output_v4float = OpTypePointer Output %v4float
%gl_FragColor = OpVariable %_ptr_Output_v4float Output
%_ptr_Input_v4float = OpTypePointer Input %v4float
)";

  const std::string before =
      R"(%main = OpFunction %void None %5
%17 = OpLabel
OpSelectionMerge %18 None
OpBranchConditional %true %19 %18
%19 = OpLabel
OpBranch %18
%18 = OpLabel
%20 = OpPhi %v4float %12 %19 %14 %17
OpStore %gl_FragColor %20
OpReturn
OpFunctionEnd
)";

  const std::string after =
      R"(%main = OpFunction %void None %5
%17 = OpLabel
OpBranch %19
%19 = OpLabel
OpBranch %18
%18 = OpLabel
OpStore %gl_FragColor %12
OpReturn
OpFunctionEnd
)";

  SinglePassRunAndCheck<opt::DeadBranchElimPass>(predefs + before,
                                                 predefs + after, true, true);
}

**Comment by greg-lunarg** [comment]
Sorry, the GLSL in the test is backwards. Here is what it should be:

// #version 140
//
// void main()
// {
// vec4 v;
// v = vec4(1.0,1.0,1.0,1.0);
// if (true)
// v = vec4(0.0,0.0,0.0,0.0);
// gl_FragColor = v;
// }

File: source/opt/dead_branch_elim_pass.cpp [hunk 9]
```diff
@@ -73,260 +72,215 @@ bool DeadBranchElimPass::GetConstInteger(uint32_t selId, uint32_t* selVal) {
...
-                caseVal = *idp;
-              } else {
-                if (caseVal == selVal) liveLabId = *idp;
-              }
+      if (GetConstInteger(terminator->GetSingleWordInOperand(0u), &selVal)) {
+        // Search switch operands for selector value, set liveLabId to
+        // corresponding label, use default if not found.
+        uint32_t icnt = 0;
+        uint32_t caseVal;
+        terminator->ForEachInOperand([
```
Comment (s-perron) [inline source/opt/dead_branch_elim_pass.cpp]: I know this was code that was already there, but, since we are doing a large change, could we change this to a for-loop.  Plus it can avoid all of the logic with `icnt`, which I always find hard to follow.

File: source/opt/dead_branch_elim_pass.cpp [hunk 10]
```diff
@@ -73,260 +72,215 @@ bool DeadBranchElimPass::GetConstInteger(uint32_t selId, uint32_t* selVal) {
...
+          ir::BasicBlock* inc =
+              get_parent_block(inst->GetSingleWordInOperand(i));
+          if (unreachableContinues.count(inc)) {
+            // Replace incoming value with undef.
+            operands.emplace_back(
+                SPV_OPERAND_TYPE_ID,
+                std::initializer_list<uint32_t>{Type2Undef(inst->type_id())});
+            operands.push_back(inst->GetInOperand(i));
+            changed = true;
+          } else if (liveBlocks.count(inc)) {
```
Comment (s-perron) [inline source/opt/dead_branch_elim_pass.cpp]: To fix the problem that Greg identified you need to check that the edge still exists, not just that the block is live.

File: source/opt/dead_branch_elim_pass.cpp (hunk 6 above)
Comment (s-perron) [reply source/opt/dead_branch_elim_pass.cpp]: I would suggest splitting it up into separate functions.  I found the function long as well.  So there would be a `FindLiveBlocks`, `MarkStructuredTargetsAsLive`, `UpdatePhis`, and `DeleteDeadBlocks`.  Or something like that.  it would make it much easier to read, at least for me.

File: source/opt/dead_branch_elim_pass.cpp [hunk 11]
```diff
@@ -73,260 +72,215 @@ bool DeadBranchElimPass::GetConstInteger(uint32_t selId, uint32_t* selVal) {
...
+      KillAllInsts(&*ebi, false);
+      // Add unreachable terminator.
+      ebi->AddInstruction(
+          MakeUnique<ir::Instruction>(context(), SpvOpUnreachable, 0, 0,
+                                      std::initializer_list<ir::Operand>{}));
+      ++ebi;
+      modified = true;
+    } else if (unreachableContinues.count(&*ebi)) {
+      // Make unreachable, but leave the label.
+      KillAllInsts(&*ebi, false);
```
Comment (s-perron) [inline source/opt/dead_branch_elim_pass.cpp]: This seems inconsistent with the handling of the Phi nodes above.  You delete the original branch, and all edges related to it, and add one back in for the header.  However, when updating the Phi node, you did not delete any edges.  You just changed it to Undef.  This is fine for the header, if the continue block branches to the header right away.  In other cases, the header will need a new entry for the Phi nodes.  For other block that are still live this will leave invalid Phi nodes.

I've assumed two things:

1) The back edge is not always in the continue block, only that it post-dominates the continue block.
2) The back edge can be a conditional branch, in the continue block, and the merge block can be reached some other way.

If these are both wrong, then there should not be a problem.

File: source/opt/dead_branch_elim_pass.cpp [hunk 12]
```diff
@@ -73,260 +72,215 @@ bool DeadBranchElimPass::GetConstInteger(uint32_t selId, uint32_t* selVal) {
...
   }
 
-  // Erase dead blocks
-  for (auto ebi = func->begin(); ebi != func->end();)
-    if (elimBlocks.find(&*ebi) != elimBlocks.end())
+  // Erase dead blocks.
+  for (auto ebi = func->begin(); ebi != func->end();) {
+    if (unreachableMerges.count(&*ebi)) {
+      // Make unreachable, but leave the label.
+      KillAllInsts(&*ebi, false);
```
Comment (s-perron) [inline source/opt/dead_branch_elim_pass.cpp]: It is hard to know what the false means here, even with the comment above.  Consider the suggestion inhttps://google.github.io/styleguide/cppguide.html#Implementation_Comments, "Function Argument Comments" section.

**Comment by alan-baker** [comment]
Thanks for the testcase Greg.

File: source/opt/dead_branch_elim_pass.cpp (hunk 11 above)
Comment (alan-baker) [reply source/opt/dead_branch_elim_pass.cpp]: If the continue is unreachable then blocks succeeding it on the backedge path are also unreachable. I guess what's missing from the phi update for the header is to add an undef to the phi from the continue instead which ever block it got a value from previously.

This is incorrect if OpLoopMerge does not imply structured control flow though (since I removed the requirement of Shader capability).

File: source/opt/dead_branch_elim_pass.cpp [hunk 13]
```diff
@@ -73,260 +72,211 @@ bool DeadBranchElimPass::GetConstInteger(uint32_t selId, uint32_t* selVal) {
 }
 
 void DeadBranchElimPass::AddBranch(uint32_t labelId, ir::BasicBlock* bp) {
+  assert(get_def_use_mgr()->GetDef(labelId) != nullptr);
   std::unique_ptr<ir::Instruction> newBranch(new ir::Instruction(
       context(), SpvOpBranch, 0, 0,
       {{spv_operand_type_t::SPV_OPERAND_TYPE_ID, {labelId}}}));
   get_def_use_mgr()->AnalyzeInstDefUse(&*newBranch);
+  context()->set_instr_block(&*newBranch, bp);
```
[the remark this answers is not in the document]
Comment (alan-baker) [reply source/opt/dead_branch_elim_pass.cpp]: I'll confirm its unnecessary and remove it.

File: source/opt/dead_branch_elim_pass.cpp (hunk 2 above)
Comment (alan-baker) [reply source/opt/dead_branch_elim_pass.cpp]: I'll add a comment.

File: source/opt/dead_branch_elim_pass.cpp [hunk 14]
```diff
@@ -73,260 +72,215 @@ bool DeadBranchElimPass::GetConstInteger(uint32_t selId, uint32_t* selVal) {
...
+  auto get_parent_block = [this](uint32_t id) {
+    return context()->get_instr_block(get_def_use_mgr()->GetDef(id));
+  };
 
-    // If constant condition/selector, replace conditional branch/switch
-    // with unconditional branch and delete merge
-    uint32_t liveLabId;
-    if (br->opcode() == SpvOpBranchConditional) {
+  // Mark live blocks reachable from the entry. Simplify constant branches and
+  // switches as you proceed to limit the number of live blocks. Be careful not
```
[the remark this answers is not in the document]
Comment (alan-baker) [reply source/opt/dead_branch_elim_pass.cpp]: Done

File: source/opt/dead_branch_elim_pass.cpp [hunk 15]
```diff
@@ -73,260 +72,215 @@ bool DeadBranchElimPass::GetConstInteger(uint32_t selId, uint32_t* selVal) {
...
+  // to eliminate backedges even if they are dead, but the header is live.
+  std::unordered_map<ir::BasicBlock*, ir::BasicBlock*> continues;
```
[the remark this answers is not in the document]
Comment (alan-baker) [reply source/opt/dead_branch_elim_pass.cpp]: Will comment

File: source/opt/dead_branch_elim_pass.cpp [hunk 16]
```diff
@@ -73,260 +72,215 @@ bool DeadBranchElimPass::GetConstInteger(uint32_t selId, uint32_t* selVal) {
...
+  bool modified = false;
+  std::unordered_set<ir::BasicBlock*> liveBlocks;
```
[the remark this answers is not in the document]
Comment (alan-baker) [reply source/opt/dead_branch_elim_pass.cpp]: I'll fix the local variables.

File: source/opt/dead_branch_elim_pass.cpp [hunk 17]
```diff
@@ -73,260 +72,215 @@ bool DeadBranchElimPass::GetConstInteger(uint32_t selId, uint32_t* selVal) {
...
+    ir::BasicBlock* block = stack.back();
+    stack.pop_back();
+
+    // Live blocks doubles as visited set.
+    if (!liveBlocks.insert(block).second) continue;
+
+    uint32_t contId = block->ContinueBlockIdIfAny();
+    if (contId != 0) continues[get_parent_block(contId)] = block;
+
+    ir::Instruction* terminator = &*block->tail();
```
[the remark this answers is not in the document]
Comment (alan-baker) [reply source/opt/dead_branch_elim_pass.cpp]: Thanks

File: source/opt/dead_branch_elim_pass.cpp [hunk 18]
```diff
@@ -73,260 +72,215 @@ bool DeadBranchElimPass::GetConstInteger(uint32_t selId, uint32_t* selVal) {
...
-    auto dbi = bi;
-    ++dbi;
-    uint32_t dLabId = (*dbi)->id();
-    while (dLabId != mergeLabId) {
-      if (!HasNonPhiNonBackedgeRef(dLabId)) {
-        // Kill use/def for all instructions and mark block for elimination
-        backedges_.erase((*dbi)->terminator());
-        KillAllInsts(*dbi);
-        elimBlocks.insert(*dbi);
+    // Don't simplify branches of continue blocks. A path from the continue to
```
[the remark this answers is not in the document]
Comment (alan-baker) [reply source/opt/dead_branch_elim_pass.cpp]: Good catch, meant to say header.

File: source/opt/dead_branch_elim_pass.cpp (hunk 5 above)
Comment (alan-baker) [reply source/opt/dead_branch_elim_pass.cpp]: Updated

File: source/opt/dead_branch_elim_pass.cpp (hunk 4 above)
Comment (alan-baker) [reply source/opt/dead_branch_elim_pass.cpp]: Ok

File: source/opt/dead_branch_elim_pass.cpp [hunk 19]
```diff
@@ -364,7 +314,8 @@ Pass::Status DeadBranchElimPass::ProcessImpl() {
   ProcessFunction pfn = [this](ir::Function* fp) {
     return EliminateDeadBranches(fp);
   };
-  bool modified = ProcessEntryPointCallTree(pfn, get_module());
+  // bool modified = ProcessEntryPointCallTree(pfn, get_module());
```
[the remark this answers is not in the document]
Comment (alan-baker) [reply source/opt/dead_branch_elim_pass.cpp]: Done

File: source/opt/dead_branch_elim_pass.cpp (hunk 6 above)
Comment (alan-baker) [reply source/opt/dead_branch_elim_pass.cpp]: I'll do some refactoring.

File: source/opt/dead_branch_elim_pass.cpp (hunk 7 above)
Comment (alan-baker) [reply source/opt/dead_branch_elim_pass.cpp]: Ok I'll do this.

File: source/opt/dead_branch_elim_pass.cpp (hunk 8 above)
Comment (alan-baker) [reply source/opt/dead_branch_elim_pass.cpp]: I'll add some more explanation here.

**Comment by s-perron** [comment]
Here is an example of a continue target that is dead.  The continue target ends with a conditional branch.  The branch goes to the header or the merge block (which is not dead).  The Phi node in the merge bloc would have to be updated.

[continue_problem.spvasm.zip](https://github.com/KhronosGroup/SPIRV-Tools/files/1612034/continue_problem.spvasm.zip)

File: source/opt/dead_branch_elim_pass.cpp [hunk 20]
```diff
@@ -73,260 +73,268 @@ bool DeadBranchElimPass::GetConstInteger(uint32_t selId, uint32_t* selVal) {
...
-      {{spv_operand_type_t::SPV_OPERAND_TYPE_ID, {condId}},
-       {spv_operand_type_t::SPV_OPERAND_TYPE_ID, {trueLabId}},
-       {spv_operand_type_t::SPV_OPERAND_TYPE_ID, {falseLabId}}}));
-  get_def_use_mgr()->AnalyzeInstDefUse(&*newBranchCond);
-  bp->AddInstruction(std::move(newBranchCond));
-}
+bool DeadBranchElimPass::MarkLiveBlocks(
+    ir::Function* func, std::unordered_set<ir::BasicBlock*>* live_blocks) {
+  // |continues| maps the continue target to its corresponding header.
+  std::unordered_map<ir::BasicBlock*, ir::BasicBlock*> continues;
```
Comment (s-perron) [inline source/opt/dead_branch_elim_pass.cpp]: Why is this a map?  Looks like a set would do.

File: source/opt/dead_branch_elim_pass.cpp [hunk 21]
```diff
@@ -73,260 +73,268 @@ bool DeadBranchElimPass::GetConstInteger(uint32_t selId, uint32_t* selVal) {
...
-      uint32_t lcnt = 0;
-      uint32_t lidx = 0;
-      uint32_t icnt = 0;
-      inst->ForEachInId([&deadPreds, &icnt, &lcnt, &lidx, this](uint32_t* idp) {
-        if (icnt % 2 == 1) {
-          if (deadPreds.find(cfg()->block(*idp)) == deadPreds.end()) {
-            ++lcnt;
-            lidx = icnt - 1;
+          // Replace the phi with either a single value or a rebuilt phi.
+          // uint32_t replId;
```
Comment (dnovillo) [inline source/opt/dead_branch_elim_pass.cpp]: Remnants of old code?

File: source/opt/dead_branch_elim_pass.cpp [hunk 22]
```diff
@@ -73,260 +73,268 @@ bool DeadBranchElimPass::GetConstInteger(uint32_t selId, uint32_t* selVal) {
...
+          //
+          // We always have type and result id operands. So this phi has a
+          // single source if are two more operands beyond those.
```
Comment (dnovillo) [inline source/opt/dead_branch_elim_pass.cpp]: s/are/there are/

File: source/opt/dead_branch_elim_pass.cpp [hunk 23]
```diff
@@ -73,260 +73,268 @@ bool DeadBranchElimPass::GetConstInteger(uint32_t selId, uint32_t* selVal) {
...
+          if (operands.size() == 4) {
+            // First input data operands is at index 2.
+            uint32_t replId = operands[2u].words[0];
+            context()->ReplaceAllUsesWith(inst->result_id(), replId);
+            iter = context()->KillInst(&*inst);
+          } else {
+            // We've rewritten the operands, so first instruct the def/use
+            // manager to forget uses that in the phi before we replace them.
```
Comment (dnovillo) [inline source/opt/dead_branch_elim_pass.cpp]: s/that//

File: include/spirv-tools/optimizer.hpp [hunk 24]
```diff
@@ -295,9 +295,8 @@ Optimizer::PassToken CreateLocalSingleBlockLoadStoreElimPass();
 // For all phi functions in merge block, replace all uses with the id
 // corresponding to the living predecessor.
 //
-// This pass only works on shaders (guaranteed to have structured control
-// flow). Note that some such branches and blocks may be left to avoid
-// creating invalid control flow. Improving this is left to future work.
+// Note that some branches and blocks may be left to avoid creating invalid
+// control flow. Improving this is left to future work.
```
[the remark this answers is not in the document]
Comment (alan-baker) [reply include/spirv-tools/optimizer.hpp]: Compute kernels can still optionally declare structured constructs. In fact, it is currently the only way to specify unroll preferences and dependency information :/ I've left a TODO to improve the handling on constant conditional branches in the continue block.

File: source/opt/basic_block.cpp [hunk 25]
```diff
@@ -106,6 +106,15 @@ void BasicBlock::ForEachSuccessorLabel(
   }
 }
 
+bool BasicBlock::IsSuccessor(const ir::BasicBlock* block) const {
+  uint32_t succId = block->id();
+  bool isSuccessor = false;
+  ForEachSuccessorLabel([&isSuccessor, succId](const uint32_t label) {
+    if (label == succId) isSuccessor = true;
```
[the remark this answers is not in the document]
Comment (alan-baker) [reply source/opt/basic_block.cpp]: ForEachSuccessor takes a function with a void return. This is a similar issue to ForEachUser in the def/use manager. We probably should have variants, but we got stuck on naming previously.

File: source/opt/dead_branch_elim_pass.cpp (hunk 21 above)
Comment (alan-baker) [reply source/opt/dead_branch_elim_pass.cpp]: Looks like.

File: source/opt/dead_branch_elim_pass.h [hunk 26]
```diff
@@ -60,34 +57,71 @@ class DeadBranchElimPass : public MemPass {
...
   // and convert to a Branch to the indicated label. Delete resulting dead
-  // blocks. Assumes only structured control flow in shader. Note some such
-  // branches and blocks may be left to avoid creating invalid control flow.
-  // TODO(greg-lunarg): Remove remaining constant conditional branches and
-  // dead blocks.
+  // blocks. Note some such branches and blocks may be left to avoid creating
+  // invalid control flow.
+  // TODO(greg-lunarg): Remove remaining constant conditional branches and dead
+  // blocks.
   bool EliminateDeadBranches(ir::Function* func);
```
[the remark this answers is not in the document]
Comment (alan-baker) [reply source/opt/dead_branch_elim_pass.h]: Yes, they should be merged, but I'd rather put that in a separate PR.

File: source/opt/dead_branch_elim_pass.cpp (hunk 22 above)
Comment (alan-baker) [reply source/opt/dead_branch_elim_pass.cpp]: Thanks

Comment (alan-baker) [reply source/opt/dead_branch_elim_pass.cpp]: Ok

File: source/opt/dead_branch_elim_pass.cpp (hunk 20 above)
Comment (alan-baker) [reply source/opt/dead_branch_elim_pass.cpp]: You're right. I think I originally made it a map assuming I'd need the mapped value, but never ended up using it.

File: source/opt/dead_branch_elim_pass.h [hunk 27]
```diff
@@ -60,34 +57,71 @@ class DeadBranchElimPass : public MemPass {
...
+      ir::Function* func,
+      const std::unordered_set<ir::BasicBlock*>& live_blocks,
+      const std::unordered_map<ir::BasicBlock*, ir::BasicBlock*>&
+          unreachable_continues);
+
+  // Erases dead blocks. Any block captured in |unreachable_merges| or
+  // |unreachable_continues| is a dead block that is required to remain due to
+  // a live merge instruction in the corresponding header. These blocks will
+  // have their instructions clobbered and will become a label and terminator.
+  // Unreachable merge blocks are terminated by OpReachable, while unreachable
```
Comment (dneto0) [inline source/opt/dead_branch_elim_pass.h]: OpUnreachable

File: source/opt/dead_branch_elim_pass.h [hunk 28]
```diff
@@ -60,34 +57,71 @@ class DeadBranchElimPass : public MemPass {
...
+  // continue blocks are terminated by an unconditional branch to the header.
+  // Otherwise, blocks are dead if not explicitly captured in |live_blocks| and
+  // are totally removed.
+  //
+  // |unreachable_continues| maps continue targets that cannot be reached to
+  // merge instruction that declares them.
```
Comment (dneto0) [inline source/opt/dead_branch_elim_pass.h]: "to the corresponding header block" rather than "to merge instruction"

File: test/opt/dead_branch_elim_test.cpp [hunk 29]
```diff
@@ -1346,6 +1346,569 @@ OpFunctionEnd
...
+OpSelectionMerge %14 None
+OpBranchConditional %false %5 %6
+%5 = OpLabel
+%10 = OpLogicalAnd %bool %true %false
+OpBranch %7
+%6 = OpLabel
+%11 = OpLogicalOr %bool %true %false
+OpBranch %7
+%7 = OpLabel
+; This phi is in a block preceeding the merge %14!
```
Comment (dneto0) [inline test/opt/dead_branch_elim_test.cpp]: Good

File: source/opt/dead_branch_elim_pass.h [hunk 30]
```diff
@@ -60,34 +57,71 @@ class DeadBranchElimPass : public MemPass {
...
+  void MarkUnreachableStructuredTargets(
+      const std::unordered_set<ir::BasicBlock*>& live_blocks,
+      std::unordered_set<ir::BasicBlock*>* unreachable_merges,
+      std::unordered_map<ir::BasicBlock*, ir::BasicBlock*>*
+          unreachable_continues);
+
+  // Fix phis in reachable blocks so that only live (or unremovable) incoming
+  // edges are present. If the block now only has a single live incoming edge,
+  // remove the phi and replace its uses with its data input.
```
[the remark this answers is not in the document]
Comment (dneto0) [reply source/opt/dead_branch_elim_pass.h]: Maybe the method comment should be updated to address this situation.  How about adding:

"If the single remaining incoming edge is from the phi itself, then the phi is in an unreachable single-block-loop.  Either the block is dead and will be removed, or it's reachable from an unreachable continue target.  In the latter case that continue target block will be collapsed into a block that only branches back to its header and we'll eliminate the block with the phi."

**Comment by greg-lunarg** [comment]
It has passed my regression testing.

**Comment by dneto0** [comment]
Rebased, squashed, and pushed into master as 1b6cfd34092ad
